### PR TITLE
Copied in tls lib from ct-go and removed project dependency

### LINKED ISSUE #2733: Drop dependency from certificate-transparency-go
Currently, there is a circular dependency between `trillian` and `certificate-transparency-go` modules:

```
~/go/src/github.com/google/trillian$ go mod graph | grep " github.com/google/certificate-transparency-go"
github.com/google/trillian github.com/google/certificate-transparency-go@v1.1.2-0.20210512142713-bed466244fa6
github.com/google/trillian@v1.3.14-0.20210511103300-67b5f349eefa github.com/google/certificate-transparency-go@v1.1.2-0.20210422104406-9f33727a7a18
github.com/google/trillian@v1.3.14-0.20210409160123-c5ea3abd4a41 github.com/google/certificate-transparency-go@v1.0.21
```

In the reverse direction:
```
~/go/src/github.com/google/certificate-transparency-go$ go mod graph | grep " github.com/google/trillian"
github.com/google/certificate-transparency-go github.com/google/trillian@v1.4.1-0.20220503161811-32257c5c3ca0
github.com/google/certificate-transparency-go@v1.1.2-0.20210512142713-bed466244fa6 github.com/google/trillian@v1.3.14-0.20210511103300-67b5f349eefa
github.com/google/certificate-transparency-go@v1.1.2-0.20210422104406-9f33727a7a18 github.com/google/trillian@v1.3.14-0.20210409160123-c5ea3abd4a41
```

The only [used](https://github.com/google/trillian/search?q=certificate-transparency-go) package from `certificate-transparency-go` repository is `tls` serialisation format imported [here](https://github.com/google/trillian/blob/ec5e4105d3539d9d42e5ea0c1b18b362608e542e/types/logroot.go#L23).

### PR DESCRIPTION
**Pull request by mhutchinson**
This fixes #2733

More changes than I'd like to go.{mod,sum} as a result of `go mod tidy`.

### CODE REVIEW

**Review by AlCutter — APPROVED** [review]

**Comment by pavelkalinnikov** [comment]
This code had tests, should they be added too?

**Comment by mhutchinson** [comment]
> This code had tests, should they be added too?

IMO it's wasted cycles running tests for code that must remain a direct copy of the tested code. There is no case where this code should not be a direct copy of the version in ct-go.

I did originally copy them across and backed out of it when I found the test referenced a greater surface area in the original repo. This meant either copying more stuff across to make the tests build, or else modifying the test file to remove cases that referenced this extra surface. Either route seemed like it was making the problem worse.

File: go.sum [hunk 1]
```diff
@@ -413,19 +319,9 @@ github.com/google/pprof v0.0.0-20210601050228-01bbb1931b22/go.mod h1:kpwsk12EmLe
 github.com/google/pprof v0.0.0-20210609004039-a478d1d731e9/go.mod h1:kpwsk12EmLew5upagYY7GY0pfYCcupk39gWOCRROcvE=
 github.com/google/pprof v0.0.0-20210720184732-4bb14d4b1be1/go.mod h1:kpwsk12EmLew5upagYY7GY0pfYCcupk39gWOCRROcvE=
 github.com/google/renameio v0.1.0/go.mod h1:KWCgfxg9yswjAJkECMjeO8J8rahYeXnNhOm40UhjYkI=
-github.com/google/rpmpack v0.0.0-20191226140753-aa36bfddb3a0/go.mod h1:RaTPr0KUf2K7fnZYLNDrr8rxAamWs3iNywJLtQ2AzBg=
-github.com/google/subcommands v1.0.1/go.mod h1:ZjhPrFU+Olkh9WazFPsl27BQ4UPiG37m3yTrtFlrHVk=
-github.com/google/trillian v1.3.14-0.20210409160123-c5ea3abd4a41/go.mod h1:1dPv0CUjNQVFEDuAUFhZql16pw/VlPgaX8qj+g5pVzQ=
-github.com/google/trillian v1.3.14-0.20210511103300-67b5f349eefa/go.mod h1:s4jO3Ai4NSvxucdvqUHON0bCqJyoya32eNw6XJwsmNc=
```
Comment (pavelkalinnikov) [inline go.sum:419]: Win!

**Review by pavelkalinnikov — COMMENTED** [review]
Tests are part of this code. Plus, having tests is a safety measure, in case the intent to have this code in sync is not fulfilled. I would argue that passing tests is one of the most worthwhile things to spend cycles on :D Far more cycles are spent on wrapping things in Docker and shipping them to some Cloud places.

Of the options that you mentioned, I would perhaps go with the refactoring approach. For example, those tests that require a larger surface could be factored out so that they wouldn't need to be copied.

An example of tests splitting: [range_test.go](https://github.com/transparency-dev/merkle/blob/main/compact/range_test.go) and [range_internal_test.go](https://github.com/transparency-dev/merkle/blob/main/compact/range_internal_test.go). This was done for a different reason though: avoiding circular dependencies.

**Pull request merged**